pkg/cmd/pulumi/packagecmd/package_publish_test.go

```diff
@@ -0,0 +1,338 @@
+			readmeContent:  "# Test README\nThis is a test readme.",
+			installContent: "# Installation\nHow to install this package.",
+		},
+		{
+			name: "error when readme is omitted",
+			args: publishPackageArgs{
+				source:    "pulumi",
+				publisher: "publisher",
+			},
+			packageSource: "testpackage",
+			packageParams: []string{},
+			mockSchema: &schema.Package{
+				Name:     "testpkg",
+				Version:  &version,
+				Provider: &schema.Resource{},
+			},
+			expectedErr: "no readme specified, please provide the path to the readme file",
+		},
+		{
+			name: "error when publish fails",
+			args: publishPackageArgs{
+				source:    "pulumi",
+				publisher: "publisher",
+			},
+			packageSource: "testpackage",
+			packageParams: []string{},
+			mockSchema: &schema.Package{
+				Name:     "testpkg",
+				Version:  &version,
+				Provider: &schema.Resource{},
+			},
+			publishErr:     errors.New("publish failed"),
+			expectedErr:    "failed to publish package",
+			readmeContent:  "# Test README\nThis is a test readme.",
+			installContent: "# Installation\nHow to install this package.",
+		},
+		{
+			name: "successful publish without installation docs",
+			args: publishPackageArgs{
+				source:    "pulumi",
+				publisher: "publisher",
+			},
+			packageSource: "testpackage",
+			packageParams: []string{},
+			mockSchema: &schema.Package{
+				Name:     "testpkg",
+				Version:  &version,
+				Provider: &schema.Resource{},
+			},
+			readmeContent: "# Test README\nThis is a test readme.",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			tempDir := t.TempDir()
+
+			var readmePath string
+			if tt.readmeContent != "" {
+				readmeFile, err := os.Create(path.Join(tempDir, "readme.md"))
+				require.NoError(t, err)
+				_, err = readmeFile.WriteString(tt.readmeContent)
+				require.NoError(t, err)
+				require.NoError(t, readmeFile.Close())
+				readmePath = readmeFile.Name()
+				tt.args.readmePath = readmePath
+			}
+
+			var installDocsPath string
+			if tt.installContent != "" {
+				installDocsFile, err := os.Create(path.Join(tempDir, "install.md"))
+				require.NoError(t, err)
+				_, err = installDocsFile.WriteString(tt.installContent)
+				require.NoError(t, err)
+				require.NoError(t, installDocsFile.Close())
+				installDocsPath = installDocsFile.Name()
+				tt.args.installDocsPath = installDocsPath
+			}
+
+			mockPackageRegistry := &backend.MockPackageRegistry{}
+			mockBackend := &backend.MockBackend{
+				GetPackageRegistryF: func() (backend.PackageRegistry, error) {
+					return mockPackageRegistry, nil
+				},
+			}
+
+			lm := &cmdBackend.MockLoginManager{
+				LoginF: func(
+					ctx context.Context,
+					ws pkgWorkspace.Context,
+					sink diag.Sink,
+					url string,
+					project *workspace.Project,
+					setCurrent bool,
+					color colors.Colorization,
+				) (backend.Backend, error) {
+					return mockBackend, nil
+				},
+			}
```

Comment:
We already have an established way to mock out backends:

https://github.com/pulumi/pulumi/blob/512006fbdccb7cae21d915fec9cfae24b4c005e6/pkg/cmd/pulumi/newcmd/new_test.go#L41-L47

Reply:
Nice, I didn't know/see that there's `backend.CurrentBackend`. Also simplifies the code of the command itself!

Reply:
There's one drawback of this approach though. By using a globally mocked BackendInstance we can no longer run the tests in parallel. They still run fast enough so I'm gonna go with it anyways